File: data-structures/stack/Stack.js
import LinkedList from '../linked-list/LinkedList.js'

class Stack {
    /** @type {LinkedList} */
    linkedList = new LinkedList()

    /**
     * @returns {boolean}
     */
    isEmpty() {
        return !this.linkedList.head
    }

    /**
     * Read the element at the front of the stack without removing it.
     * @returns {*}
     */
    peek() {
        return this.linkedList.head?.value
    }

    /**
     * Add a new element to the front of the stack (the head of the linked list).
     * This element will be processed first.
     * @param {*} value
     */
    push(value) {
        this.linkedList.prepend(value)
    }

    /**
     * Remove and return the element at the front of the queue (the head of the linked list).
     * @returns {*}
     */
    pop() {
        const deletedHead = this.linkedList.deleteHead()
        return deletedHead?.value
    }
}

export default Stack

File: data-structures/stack/stack.go
package stack

import (
	"container/list"
)

type Stack struct {
	list *list.List
}

func New() *Stack {
	return &Stack{
		list: list.New(),
	}
}

func (s *Stack) IsEmpty() bool {
	return s.list.Front() == nil
}

func (s *Stack) Push(value interface{}) {
	s.list.PushFront(value)
}

func (s *Stack) Pop() interface{} {
	if el := s.list.Front(); el != nil {
		s.list.Remove(el)
		return el.Value
	}
	return nil
}

func (s *Stack) Peek() interface{} {
	if el := s.list.Front(); el != nil {
		return el.Value
	}
	return nil
}
